File: leetcode/201-1000/leetcode235.cpp
#include <iostream>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x)
        : val(x)
        , left(NULL)
        , right(NULL)
    {
    }
};

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q)
    {
        TreeNode* now;
        if (q->val < p->val) {
            now = q;
            q = p;
            p = now;
        }
        now = root;
        while (now != nullptr) {
            if (now->val >= p->val && now->val <= q->val) {
                return now;
            } else if (now->val <= p->val) {
                now = now->right;
            } else {
                now = now->left;
            }
        }
        return now;
    }
};
int main(void)
{
    return 0;
}

File: leetcode/100-999/leetcode235.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def lowestCommonAncestor(
        self, root: "TreeNode", p: "TreeNode", q: "TreeNode"
    ) -> "TreeNode":
        p_parent, q_parent = [], []

        def chaewon(root: TreeNode, parent: list, target: TreeNode):
            if not root:
                return False
            parent.append(root)
            if root == target:
                return True
            if chaewon(root=root.left, parent=parent, target=target):
                return True
            if chaewon(root=root.right, parent=parent, target=target):
                return True
            parent.pop()
            return False

        chaewon(root=root, parent=p_parent, target=p)
        chaewon(root=root, parent=q_parent, target=q)

        res = root
        for i in range(min(len(p_parent), len(q_parent))):
            if p_parent[i] == q_parent[i]:
                res = p_parent[i]
            else:
                break
        return res
